src/cxx_support.rs

```diff
@@ -0,0 +1,214 @@
+/// had its constructor called.
+#[cfg(feature = "cxx")]
+pub struct MaybeUninitUniquePtr<T: UniquePtrTarget>(*mut T);
+
+impl<T: UniquePtrTarget> MaybeUninitUniquePtr<T> {
+  /// Declare that a pointer which we have received from C++
+  /// is, in fact, not initialized but is instead just a blank
+  /// storage area yet to have its constructor called.
+  ///
+  /// Safety: destroying this type results in a memory leak. Callers
+  /// need to take care to either convert this to a normal, fully-
+  /// initialized [`UniquePtr`] by calling [`Self::assume_init`] or
+  /// to manually take care to deallocate the underlying storage.
+  pub unsafe fn from_ptr(ptr: *mut T) -> Self {
+    assert!(!ptr.is_null());
+    Self(ptr)
+  }
+
+  /// Get a pointer to the storage within this uninitialized space.
+  /// This can be used to initialize it.
+  ///
+  /// Safety: callers must call this exactly once per instance, or
+  /// the data may be double-initialized.
+  pub unsafe fn get_uninitialized_space(&mut self) -> Pin<&mut T> {
+    Pin::new_unchecked(self.0.as_mut().unwrap())
+  }
+
+  /// Declare that this [`UniquePtr`] is initialized and has had its
+  /// constructor called.
+  ///
+  /// Safety: callers must ensure that the data was initialized exactly
+  /// once before calling this function.
+  pub unsafe fn assume_init(self) -> UniquePtr<T> {
+    UniquePtr::from_raw(self.0)
+  }
+
+  /// Return the underlying raw pointer such that it can be freed.
+  ///
+  /// Safety: callers must ensure they deallocate the resulting pointer
+  /// to avoid a memory leak. The normal purpose of this method is to
+  /// return a pointer which has been allocated, but not initialized;
+  /// in such a case callers would need to free the memory without calling
+  /// the destructor. However this is not enforced - callers will need
+  /// to free this memory appropriately depending on the lifecycle of this
+  /// object.
+  pub unsafe fn get_for_freeing(self) -> *mut T {
+    self.0
+  }
+}
+
+/// A type which has the ability to create a cxx::UniquePtr which points to
+/// an area of memory with the correct size and alignment, but has not yet had
+/// its constructor called.
+#[cfg(feature = "cxx")]
+pub trait MakeUninitUniquePtr
+where
+  Self: UniquePtrTarget + Sized,
+{
+  // Make a UniquePtr which is allocated (i.e. is not null) but has not yet
+  // had a constructor called.
+  fn make_uninit_unique_ptr() -> MaybeUninitUniquePtr<Self>;
+  // Free a UniquePtr which is allocated (i.e. is not null) but has not yet
+  // had a constructor called.
+  fn free_uninit_unique_ptr(ptr: MaybeUninitUniquePtr<Self>);
+}
+
+impl<T: MakeUninitUniquePtr> Emplace<T> for UniquePtr<T> {
+  fn try_emplace<N: TryNew<Output = T>>(n: N) -> Result<Pin<Self>, N::Error> {
+    unsafe {
+      let mut uninit_ptr = T::make_uninit_unique_ptr();
+      let uninit = uninit_ptr.get_uninitialized_space();
+      // TODO - see if we avoid this transmute.
```

Comment:
I think this would require `try_new` to accept a trait instead of a `MaybeUninit`. Say,

```rust
trait SomethingWotMightNotBeInitiailized {
   fn assume_init(); // and the rest of the MaybeUninit API
}

impl SomethingWotMightNotBeInitiailized for MaybeUninit { ... }

impl SomethingWotMightNotBeInitiailized for MaybeUninitUniquePtr { ... }
```

Reply:
If this function returns a Pin<&mut MaybeUninit<T>> I don't think there should be an issue...